clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AbstractAvroStoreClient.java

```diff
@@ -248,6 +250,30 @@ protected String getComputeRequestPath() {
     return TYPE_COMPUTE + "/" + storeName;
   }
 
+  /**
+   * This function will try to initialize the store client at most once triggered by request.
+   * If the store initialization fails even with the internal retries, all the subsequent requests will fail,
+   * and the reason behind this design is that we don't want the retried store init occupy the application
+   * thread too long for every request, so that application's fallback logic can kick in.
+   */
+  protected RecordSerializer<K> getKeySerializerForRequest() {
+    if (keySerializer != null) {
+      return keySerializer;
+    }
+    if (whetherStoreInitTriggeredByRequestFail) {
+      // Store init already fails.
+      throw new VeniceClientException("Failed to init store client for store: " + storeName);
+    }
+    synchronized (this) {
```

Comment:
I wonder whether we can actually acquire the lock here and fail fast. IIUC, the background thread would call `getKeySerializerWithRetryWithLongInterval` and keep the lock for 10 seconds.

Reply:
I don't follow this comment.
The async warmup will try to do store init as a best effort and if the async warmup can be done because d2 client is not ready, this request triggered store init won't work either.
So technically, there is no difference.
The assumption here is that the internal d2 client must be ready when sending out the first request and if this won't work for some application, the application needs to modify their logic, otherwise, the first query will always fail no matter what.